File: tests/csharp/tests-csharp-nunit-allure3-selenium/Helpers/LocalChromePin.cs
using System.Text.RegularExpressions;
using Config;

namespace Helpers;

/// <summary>
/// Pins local Chrome to the Chrome for Testing build in <c>chrome-for-testing.properties</c>.
/// Bypasses Selenium Manager so system Chrome is never used silently.
/// </summary>
public static class LocalChromePin
{
    private const string PinFile = "chrome-for-testing.properties";
    private const string Installer = "scripts/install-chrome-for-testing.sh";

    public sealed record Binaries(string Chrome, string Driver);

    public static Binaries Apply(string? browserVersion) => Resolve(browserVersion);

    public static Binaries Resolve(string? browserVersion)
    {
        if (string.IsNullOrWhiteSpace(browserVersion))
        {
            throw new InvalidOperationException(
                "browserVersion is required for local Chrome (canon: 148). "
                + "Do not run e2e on system Chrome without explicit override.");
        }

        var version = PinnedVersion();
        RequireSameMajor(browserVersion, version);

        var chrome = ExecutableOverride("CHROME_BINARY_PATH") ?? ChromeBinary(version);
        if (!IsExecutable(chrome))
        {
            throw NotInstalled($"Chrome {version} browser binary", chrome);
        }

        var driver = ExecutableOverride("CHROMEDRIVER_PATH") ?? ChromeDriverPath(version);
        if (!IsExecutable(driver))
        {
            var cached = SeleniumCacheDriver(version);
            if (!IsExecutable(cached))
            {
                throw NotInstalled($"chromedriver for Chrome {version}", driver);
            }

            driver = cached;
        }

        return new Binaries(chrome, driver);
    }

    public static string PinnedVersion()
    {
        var overrideVersion = Environment.GetEnvironmentVariable("CHROME_FOR_TESTING_VERSION")?.Trim();
        if (!string.IsNullOrWhiteSpace(overrideVersion))
        {
            return overrideVersion;
        }

        var path = PinPath();
        if (!File.Exists(path))
        {
            throw new InvalidOperationException($"{PinFile} is missing from the tests module");
        }

        foreach (var line in File.ReadAllLines(path))
        {
            var trimmed = line.Trim();
            if (!trimmed.StartsWith("version=", StringComparison.Ordinal))
            {
                continue;
            }

            var version = trimmed["version=".Length..].Trim();
            if (version.Length > 0)
            {
                return version;
            }
        }

        throw new InvalidOperationException($"No version= entry in {PinFile}");
    }

    private static string PinPath()
    {
        var fromOutput = Path.Combine(AppContext.BaseDirectory, PinFile);
        if (File.Exists(fromOutput))
        {
            return fromOutput;
        }

        return Path.Combine(Config.ConfigFiles.ModuleDir(), PinFile);
    }

    private static string? ExecutableOverride(string environmentVariable)
    {
        var value = Environment.GetEnvironmentVariable(environmentVariable);
        return string.IsNullOrWhiteSpace(value) ? null : value.Trim();
    }

    private static void RequireSameMajor(string browserVersion, string pinnedVersion)
    {
        var requested = Major(browserVersion);
        var pinned = Major(pinnedVersion);
        if (requested != pinned)
        {
            throw new InvalidOperationException(
                $"browserVersion={browserVersion} asks for Chrome {requested}, but the pinned build is {pinnedVersion}. "
                + $"Align them: bump version= in {PinFile}, or set browserVersion to {pinned}.");
        }
    }

    private static string Major(string version) => version.Split('.')[0];

    private static string ChromeForTestingRoot()
    {
        var overridePath = Environment.GetEnvironmentVariable("CHROME_FOR_TESTING_PATH");
        if (!string.IsNullOrWhiteSpace(overridePath))
        {
            return overridePath.Trim();
        }

        return Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
            ".local", "share", "chrome-for-testing");
    }

    private static string PlatformDir()
    {
        if (OperatingSystem.IsMacOS())
        {
            return System.Runtime.InteropServices.RuntimeInformation.ProcessArchitecture
                is System.Runtime.InteropServices.Architecture.Arm64
                ? "mac_arm"
                : "mac";
        }

        if (OperatingSystem.IsLinux())
        {
            return "linux";
        }

        throw new InvalidOperationException($"Unsupported OS for LocalChromePin: {Environment.OSVersion}");
    }

    private static string SeleniumCacheArch() => PlatformDir() switch
    {
        "mac_arm" => "mac-arm64",
        "mac" => "mac-x64",
        "linux" => "linux64",
        _ => throw new InvalidOperationException($"Unsupported platform: {PlatformDir()}"),
    };

    private static string ChromeBinary(string version)
    {
        var versionDir = Path.Combine(ChromeForTestingRoot(), "chrome", $"{PlatformDir()}-{version}");
        return PlatformDir() switch
        {
            "mac_arm" => Path.Combine(versionDir, "chrome-mac-arm64", "Google Chrome for Testing.app", "Contents", "MacOS", "Google Chrome for Testing"),
            "mac" => Path.Combine(versionDir, "chrome-mac-x64", "Google Chrome for Testing.app", "Contents", "MacOS", "Google Chrome for Testing"),
            "linux" => Path.Combine(versionDir, "chrome-linux64", "chrome"),
            _ => throw new InvalidOperationException($"Unsupported platform: {PlatformDir()}"),
        };
    }

    private static string ChromeDriverPath(string version)
    {
        var versionDir = Path.Combine(ChromeForTestingRoot(), "chromedriver", $"{PlatformDir()}-{version}");
        return PlatformDir() switch
        {
            "mac_arm" => Path.Combine(versionDir, "chromedriver-mac-arm64", "chromedriver"),
            "mac" => Path.Combine(versionDir, "chromedriver-mac-x64", "chromedriver"),
            "linux" => Path.Combine(versionDir, "chromedriver-linux64", "chromedriver"),
            _ => throw new InvalidOperationException($"Unsupported platform: {PlatformDir()}"),
        };
    }

    private static string SeleniumCacheDriver(string version) =>
        Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
            ".cache", "selenium", "chromedriver",
            SeleniumCacheArch(),
            version,
            "chromedriver");

    private static bool IsExecutable(string path) => File.Exists(path);

    private static InvalidOperationException NotInstalled(string what, string expected) =>
        new(
            $"{what} not found at {expected}. "
            + $"Install the pinned build (not system Chrome), from the tests module root:\n  {Installer}");
}

File: tests/rust/tests-rust-testing-reqwest-selenium/src/local_chrome_pin.rs
use std::path::{Path, PathBuf};

use crate::config::module_dir;

const PIN_FILE: &str = "chrome-for-testing.properties";
const INSTALLER: &str = "scripts/install-chrome-for-testing.sh";

#[derive(Debug, Clone)]
pub struct ChromeBinaries {
    pub chrome: PathBuf,
    pub driver: PathBuf,
}

/// Pins local Chrome to the Chrome for Testing build in `chrome-for-testing.properties`.
pub fn apply(browser_version: &str) -> Result<ChromeBinaries, String> {
    resolve(browser_version)
}

pub fn resolve(browser_version: &str) -> Result<ChromeBinaries, String> {
    if browser_version.trim().is_empty() {
        return Err(
            "browserVersion is required for local Chrome (canon: 148). \
             Do not run e2e on system Chrome without explicit override."
                .into(),
        );
    }
    let version = pinned_version()?;
    require_same_major(browser_version, &version)?;

    let chrome = executable_override("CHROME_BINARY_PATH").unwrap_or_else(|| chrome_binary(&version));
    if !is_executable(&chrome) {
        return Err(not_installed(
            &format!("Chrome {version} browser binary"),
            &chrome,
        ));
    }

    let mut driver =
        executable_override("CHROMEDRIVER_PATH").unwrap_or_else(|| chromedriver_path(&version));
    if !is_executable(&driver) {
        let cached = selenium_cache_driver(&version);
        if !is_executable(&cached) {
            return Err(not_installed(
                &format!("chromedriver for Chrome {version}"),
                &driver,
            ));
        }
        driver = cached;
    }

    Ok(ChromeBinaries { chrome, driver })
}

pub fn pinned_version() -> Result<String, String> {
    let override_version = std::env::var("CHROME_FOR_TESTING_VERSION").unwrap_or_default();
    if !override_version.trim().is_empty() {
        return Ok(override_version.trim().to_string());
    }
    let path = pin_path();
    if !path.is_file() {
        return Err(format!("{PIN_FILE} is missing from the tests module"));
    }
    let text = std::fs::read_to_string(&path).map_err(|err| err.to_string())?;
    for line in text.lines() {
        let trimmed = line.trim();
        if let Some(version) = trimmed.strip_prefix("version=") {
            let version = version.trim();
            if !version.is_empty() {
                return Ok(version.to_string());
            }
        }
    }
    Err(format!("No version= entry in {PIN_FILE}"))
}

fn pin_path() -> PathBuf {
    module_dir().join(PIN_FILE)
}

fn executable_override(environment_variable: &str) -> Option<PathBuf> {
    let value = std::env::var(environment_variable).ok()?;
    let trimmed = value.trim();
    if trimmed.is_empty() {
        None
    } else {
        Some(PathBuf::from(trimmed))
    }
}

fn require_same_major(browser_version: &str, pinned_version: &str) -> Result<(), String> {
    let requested = major(browser_version);
    let pinned = major(pinned_version);
    if requested != pinned {
        return Err(format!(
            "browserVersion={browser_version} asks for Chrome {requested}, but the pinned build is {pinned_version}. \
             Align them: bump version= in {PIN_FILE}, or set browserVersion to {pinned}."
        ));
    }
    Ok(())
}

pub fn major(version: &str) -> String {
    version.split('.').next().unwrap_or(version).to_string()
}

fn chrome_for_testing_root() -> PathBuf {
    if let Ok(override_path) = std::env::var("CHROME_FOR_TESTING_PATH") {
        let trimmed = override_path.trim();
        if !trimmed.is_empty() {
            return PathBuf::from(trimmed);
        }
    }
    home_dir().join(".local/share/chrome-for-testing")
}

fn home_dir() -> PathBuf {
    std::env::var("HOME")
        .or_else(|_| std::env::var("USERPROFILE"))
        .map(PathBuf::from)
        .unwrap_or_else(|_| PathBuf::from("."))
}

fn platform_dir() -> Result<&'static str, String> {
    match std::env::consts::OS {
        "macos" => {
            if std::env::consts::ARCH == "aarch64" {
                Ok("mac_arm")
            } else {
                Ok("mac")
            }
        }
        "linux" => Ok("linux"),
        other => Err(format!("Unsupported OS for LocalChromePin: {other}")),
    }
}

fn selenium_cache_arch(platform: &str) -> Result<&'static str, String> {
    match platform {
        "mac_arm" => Ok("mac-arm64"),
        "mac" => Ok("mac-x64"),
        "linux" => Ok("linux64"),
        other => Err(format!("Unsupported platform: {other}")),
    }
}

fn chrome_binary(version: &str) -> PathBuf {
    let platform = platform_dir().unwrap_or("linux");
    let version_dir = chrome_for_testing_root().join("chrome").join(format!("{platform}-{version}"));
    match platform {
        "mac_arm" => version_dir.join(
            "chrome-mac-arm64/Google Chrome for Testing.app/Contents/MacOS/Google Chrome for Testing",
        ),
        "mac" => version_dir.join(
            "chrome-mac-x64/Google Chrome for Testing.app/Contents/MacOS/Google Chrome for Testing",
        ),
        _ => version_dir.join("chrome-linux64/chrome"),
    }
}

fn chromedriver_path(version: &str) -> PathBuf {
    let platform = platform_dir().unwrap_or("linux");
    let version_dir = chrome_for_testing_root()
        .join("chromedriver")
        .join(format!("{platform}-{version}"));
    match platform {
        "mac_arm" => version_dir.join("chromedriver-mac-arm64/chromedriver"),
        "mac" => version_dir.join("chromedriver-mac-x64/chromedriver"),
        _ => version_dir.join("chromedriver-linux64/chromedriver"),
    }
}

fn selenium_cache_driver(version: &str) -> PathBuf {
    let platform = platform_dir().unwrap_or("linux");
    let arch = selenium_cache_arch(platform).unwrap_or("linux64");
    home_dir()
        .join(".cache/selenium/chromedriver")
        .join(arch)
        .join(version)
        .join("chromedriver")
}

fn is_executable(path: &Path) -> bool {
    path.is_file()
}

fn not_installed(what: &str, expected: &Path) -> String {
    format!(
        "{what} not found at {}. Install the pinned build (not system Chrome), from the tests module root:\n  {INSTALLER}",
        expected.display()
    )
}
